apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabOptimize/Components/AllowChar.tsx

```diff
@@ -0,0 +1,111 @@
+import { allArtifactSlotKeys, charKeyToLocCharKey, LocationCharacterKey, LocationKey } from "@genshin-optimizer/consts";
+import SettingsAccessibilityIcon from '@mui/icons-material/SettingsAccessibility';
+import { Box, Button, CardActionArea, CardContent, Divider, Grid, Typography } from "@mui/material";
+import { useCallback, useContext, useDeferredValue, useEffect, useMemo } from "react";
+import { Trans, useTranslation } from "react-i18next";
+import SlotIcon from "../../../../../Components/Artifact/SlotIcon";
+import CardDark from "../../../../../Components/Card/CardDark";
+import CardLight from "../../../../../Components/Card/CardLight";
+import CharacterCardPico from "../../../../../Components/Character/CharacterCardPico";
+import CloseButton from "../../../../../Components/CloseButton";
+import ModalWrapper from "../../../../../Components/ModalWrapper";
+import SqBadge from "../../../../../Components/SqBadge";
+import { CharacterContext } from "../../../../../Context/CharacterContext";
+import { DatabaseContext } from "../../../../../Database/Database";
+import useBoolState from "../../../../../ReactHooks/useBoolState";
+import useForceUpdate from "../../../../../ReactHooks/useForceUpdate";
+import { toggleArr } from "../../../../../Util/Util";
+import useBuildSetting from "../useBuildSetting";
+export default function AllowChar({ disabled = false, numArtsEquippedUsed }: { disabled?: boolean, numArtsEquippedUsed: number }) {
+  const { t } = useTranslation("page_character_optimize")
+  const { character: { key: characterKey } } = useContext(CharacterContext)
+  const { buildSetting: { allowLocations }, buildSettingDispatch } = useBuildSetting(characterKey)
+  const { database } = useContext(DatabaseContext)
+  const [show, onOpen, onClose] = useBoolState(false)
+  const [dbDirty, forceUpdate] = useForceUpdate()
+  const deferredDbDirty = useDeferredValue(dbDirty)
+
+  useEffect(() => database.charMeta.followAny(s => forceUpdate()), [forceUpdate, database])
+  useEffect(() => database.chars.followAny(s => forceUpdate()), [forceUpdate, database])
+
+  const locList = useMemo(() => deferredDbDirty && Array.from(new Set(
```

Comment:
Why do we need to convert to set then to array? feels like database.chars.data shouldn't have any duplicates right?

Reply:
Because converting CharKey to LocKey will give you multiple Travelers